---
title: "Covid"
author: "Daniela Corbetta"
date: "2025-01-22"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Code to reproduce results of section 5 of paper "Conformal inference for cell type annotation with graph-structured constraints"

```{r, message=FALSE}
library(SingleCellExperiment)
library(scater)
library(scran)
library(VGAM)
library(igraph)
library(ontoProc)
library(scConform)
library(BiocParallel)
```


```{r, message=FALSE}
`%notin%` = Negate(`%in%`)

cl <- getOnto("cellOnto", "2023")
```

# Load data

```{r}
sce_Covid_Bcells <- readRDS("sce_Covid_Bcells.rds")
#colnames(colData(sce_Covid_Bcells))
length(table(sce_Covid_Bcells$donor_id))
```

# Choose samples and build training/test dataset
```{r}
table(sce_Covid_Bcells$Site)
B_Ncl <- sce_Covid_Bcells[,sce_Covid_Bcells$Site=='Ncl']
barplot(prop.table(table(B_Ncl$cell_type_curated)))
```

```{r}
table(B_Ncl$cell_type_curated)
table(B_Ncl$disease)
table(B_Ncl$Status)
```




```{r}
test <- sce_Covid_Bcells[,sce_Covid_Bcells$donor_id=="MH9143273" & 
                           sce_Covid_Bcells$cell_type_curated!="B cell"]
other <- sce_Covid_Bcells[,sce_Covid_Bcells$donor_id!="MH9143273"& 
                           sce_Covid_Bcells$cell_type_curated!="B cell" &
                            sce_Covid_Bcells$Site=="Ncl" &
                            sce_Covid_Bcells$disease=="COVID-19"]

barplot(rbind(prop.table(table(other$cell_type_curated)),
              prop.table(table(test$cell_type_curated))), 
              beside=TRUE,
        legend=c("test", "train"))
length(unique(other$donor_id))
other
test
```

# Preprocess datasets

```{r}
# Add quality control metrics and filter low-quality cells
hist(other$detected, nclass = 50) # number of different features
abline(v = quantile(other$detected, probs = 0.05), col = 2)
# delete cells with less than 0.05 quantile of expressed features
low_detection <- (other$detected < quantile(other$detected, probs = 0.05)) 

hist(other$sum, nclass = 50) # total number of counts
abline(v = quantile(other$sum, probs = c(0.95)), col = 2) # quantile 0.95
# delete cells with more than quantile(other$sum, probs = c(0.95)) total counts, might be doublets
high_counts <- other$sum > quantile(other$sum, probs = c(0.95))

other <- other[, !low_detection & !high_counts]
other #32431 
```

```{r}
hist(test$detected, nclass = 50) # number of different features
abline(v = quantile(test$detected, probs = 0.05), col = 2)
# delete cells with less than 0.05 quantile of expressed features
low_detection <- (test$detected < quantile(test$detected, probs = 0.05)) 

hist(test$sum, nclass = 50) # total number of counts
abline(v = quantile(test$sum, probs = c(0.95)), col = 2) # quantile 0.95
# delete cells with more than quantile(test$sum, probs = c(0.95)) total counts, might be doublets
high_counts <- test$sum > quantile(test$sum, probs = c(0.95))

test <- test[, !low_detection & !high_counts]
test #1762
```

# Build ontology

```{r}
tags <- unique(other$cell_type_ontology_term_id)
p <- onto_plot2(cl, tags)

onto <- graph_from_graphnel(p)

## Delete instances from other ontologies
sel_ver <- V(onto)$name[c(grep("CARO", V(onto)$name), grep("BFO", V(onto)$name))]
onto <- onto - sel_ver

## Rename vertex to match annotations
V(onto)$name[V(onto)$name=="IgA\nplasma cell\nCL:0000987"] <- "IgA plasma cell"
V(onto)$name[V(onto)$name=="IgM\nplasma cell\nCL:0000986"] <- "IgM plasma cell"
V(onto)$name[V(onto)$name=="naive\nB cell\nCL:0000788"] <- "naive B cell"
V(onto)$name[V(onto)$name=="immature\nB cell\nCL:0000816"] <- "immature B cell"
V(onto)$name[V(onto)$name=="class\nswitched memory B cell\nCL:0000972"] <- "class switched memory B cell"
V(onto)$name[V(onto)$name=="unswitched\nmemory B cell\nCL:0000970"] <- "unswitched memory B cell"
V(onto)$name[V(onto)$name=="IgG\nplasma cell\nCL:0000985"] <- "IgG plasma cell"
V(onto)$name[V(onto)$name=="plasmablast\nCL:0000980"] <- "plasmablast"


pan.gr <- as_graphnel(onto)
plot(pan.gr, attrs=list(node=list(fontsize=27, shape="box")))
V(onto)$name <- gsub("\n", " ", V(onto)$name)
V(onto)$name <- sub("CL:.*", "", V(onto)$name)

pan.gr1 <- as_graphnel(onto)
plot(pan.gr1, attrs=list(node=list(fontsize=27, shape="box")))
```

# Uncertainty quantification

Build a balanced train+cal dataset by downsampling the dataset "other"
```{r}
num <- min(table(other$cell_type_curated))
cl.types <- unique(other$cell_type_curated)
set.seed(1130)
idx <- NULL
for(i in cl.types){
  cat <- which(other$cell_type_curated==i)
  idx_cat <- sample(cat, size = num, replace = FALSE)
  idx <- c(idx, idx_cat)
}
other

other.bal <- other[,idx]
barplot(prop.table(table(other$cell_type_curated)))
barplot(prop.table(table(other.bal$cell_type_curated)))
```

Fit model (Isabella N Grabski, Rafael A Irizarry, A probabilistic gene expression barcode for annotation of cell types from single-cell RNA-seq data, Biostatistics, Volume 23, Issue 4, October 2022, Pages 1150â€“1164)

```{r}
# Grabsky Irizzarry
source('cell_type_identification4_clean.R')
library(org.Hs.eg.db)

### Preparation for symbol->ENSEMBL conversion
Hs_symbol <- org.Hs.egSYMBOL
mapped_Hs_genes.symbol <- mappedkeys(Hs_symbol)
Hs_symbol.df <- as.data.frame(Hs_symbol[mapped_Hs_genes.symbol])
Hs_ensembl <- org.Hs.egENSEMBL
mapped_Hs_genes.ensembl <- mappedkeys(Hs_ensembl)
Hs_ensembl.df <- as.data.frame(Hs_ensembl[mapped_Hs_genes.ensembl])
Hs_mapping <- merge(Hs_symbol.df,Hs_ensembl.df)

rownames(spe_train) <- Hs_mapping$ensembl_id[match(rownames(spe_train),
                                                       Hs_mapping$symbol)]
rownames(test) <- Hs_mapping$ensembl_id[match(rownames(test),
                                                       Hs_mapping$symbol)]
rownames(spe_cal) <- Hs_mapping$ensembl_id[match(rownames(spe_cal),
                                                       Hs_mapping$symbol)]

# Train on reference
train.d <- trainAllReference(as.matrix(counts(spe_train)),
  spe_train$cell_type_curated,discrim_only=TRUE)

test.pred <- classifyTarget(as.matrix(counts(test)), train.d, return.probs = T)
cal.pred <- classifyTarget(as.matrix(counts(spe_cal)), train.d, return.probs = T)
```

## Conformal inference

Compare coverage with and without resampling correction
```{r}
labels <- V(onto)$name[degree(onto, mode = "out") == 0]
spe_cal <- other.bal[,cal]
# Create corresponding colData
for (i in labels) {
    colData(spe_cal)[[i]] <- cal.pred[, i]
    colData(test)[[i]] <- test.pred[, i]
}

test <- getPredictionSets(x_query=test, x_cal=spe_cal, 
                          y_cal=spe_cal$cell_type_curated,
                          alpha=0.1, follow_ontology = FALSE, resample=FALSE,
                          onto=onto, pr_name = "pr_conf_nores")

mean(mapply(`%in%`, test$cell_type_curated, test$pr_conf_nores))
# 0.976

test <- getPredictionSets(x_query=test, x_cal=spe_cal, 
                          y_cal=spe_cal$cell_type_curated,
                          alpha=0.1, follow_ontology = TRUE, resample=FALSE,
                          onto=onto, pr_name = "pr_crc_nores",
                          BPPARAM = MulticoreParam(workers = 6))
mean(mapply(`%in%`, test$cell_type_curated, test$pr_crc_nores))
# 0.9659478

test <- getPredictionSets(x_query=test, x_cal=spe_cal, 
                          y_cal=spe_cal$cell_type_curated,
                          alpha=0.1, follow_ontology = FALSE,
                          resample=TRUE,
                          onto=onto, pr_name = "pr_conf_res")

mean(mapply(`%in%`, test$cell_type_curated, test$pr_conf_res))
# 0.930

test <- getPredictionSets(x_query=test, x_cal=spe_cal, 
                          y_cal=spe_cal$cell_type_curated,
                          alpha=0.1, follow_ontology = TRUE,
                          resample=TRUE,
                          onto=onto, pr_name = "pr_crc_res1",
                          BPPARAM = MulticoreParam(workers = 6))
mean(mapply(`%in%`, test$cell_type_curated, test$pr_crc_res1))
# 0.915
```

Compare results with an estimated oracle correction that resamples based on the real values of the proportions in the test set
```{r}
resample.oracle <- function(cal, test, cal.pred, seed = NA) {
  if (!is.na(seed)) set.seed(seed)
  
  cal_freq <- prop.table(table(cal$Y))
  test_freq <- prop.table(table(test$Y))
  des_freq <- round(test_freq * length(cal$Y))
  
  idx <- NULL
  for (i in unique(test$Y)) {
    cat <- which(cal$Y == i)
    if (!is.na(des_freq[i])) {
      idx_cat <- sample(cat, size = des_freq[i], replace = TRUE)
      idx <- c(idx, idx_cat)
    }
  }
  
  return(list(cal = cal[idx, ], p.cal = cal.pred[idx, ]))
}

cal_oracle <- resample.oracle(df.cal, df.test, pr.cal, seed=1)

oracle_sets <- getPredictionSets(x_query=pr, x_cal=cal_oracle$p.cal, 
                          y_cal=cal_oracle$cal$Y,
                          alpha=0.1, follow_ontology = FALSE, resample=FALSE,
                          return_sc=FALSE, onto=onto)
mean(mapply(`%in%`, test$cell_type_curated, oracle_sets))
#0.918
oracle_sets_gr <- getPredictionSets(x_query=pr, x_cal=cal_oracle$p.cal, 
                          y_cal=cal_oracle$cal$Y,
                          alpha=0.1, follow_ontology = TRUE, resample=FALSE,
                          onto=onto, return_sc=FALSE,
                          BPPARAM = MulticoreParam(workers = 6))

mean(mapply(`%in%`, test$cell_type_curated, oracle_sets_gr))
#0.896
```

Obtain prediction sets
```{r}
cal.pred <- classifyTarget(as.matrix(counts(spe_cal)), train.d, return.probs = T)
pr_GI <- getPredictionSets(x_query=test.pred, x_cal=cal.pred, 
                          y_cal=spe_cal$cell_type_curated,
                          alpha=0.1, follow_ontology = TRUE, resample=TRUE,
                          onto=onto, return_sc = FALSE,
                          BPPARAM = MulticoreParam(workers = 6))
mean(sapply(pr_GI, length))
table(sapply(pr_GI, length))
```

```{r}
getCommAnc <- function (pred_set, onto) 
{
    if (length(pred_set)==1) 
      return(pred_set)
    else{
      com_anc <- Reduce(intersect, lapply(pred_set, function(node) {
          scConform:::.ancestors(node, onto)
      }))
      root <- V(onto)$name[degree(onto, mode = "in") == 0]
      first_anc <- com_anc[which.max(distances(onto, v = com_anc, 
          to = root))]
      return(first_anc)
    }
}
```




```{r}
test$comm_anc_GI <- vapply(
            test$pr_crc_res,
            function(x) getCommAnc(x, onto),
            character(1)
        )

## Check 
ch <- rep(NA, length(test$comm_anc_GI))
for(i in seq_along(test$comm_anc_GI))
  ch[i] <- length(scConform:::.children(test$comm_anc_GI[[i]], onto)) == length(test$pr_crc_res[[i]])
sum(!ch)

# Define the color vector
color_vector <- c(
  "IgA plasma cell" = "#1f77b4",              # Blue
  "mature B cell " = "#ff7f0e",               # Orange
  "long lived plasma cell " = "#2ca02c",      # Green
  "B cell, CD19-positive " = "#d62728",       # Red
  "memory B cell " = "#9467bd",               # Purple
  "IgM plasma cell" = "#8c564b",              # Brown
  "class switched memory B cell" = "#e377c2", # Pink
  "plasmablast" = "#7f7f7f",                  # Gray
  "lymphocyte of B lineage " = "#bcbd22",     # Olive
  "antibody secreting cell " = "#17becf",     # Cyan
  "unswitched memory B cell" = "#f7b6d2",     # Light Pink
  "immature B cell" = "#c5b0d5",              # Light Purple
  "antibody secreting cell, mature B cell" = "#ff9896", # Light Red
  "IgG plasma cell" = "gold",             # Light Blue
  "naive B cell" = "#1ABC9C"                  # Light Gray
)

# Add point predictions
test$pr_label <- pr.class

test <- logNormCounts(test)
test <- runPCA(test, name="PCA")
set.seed(1)
test <- runTSNE(test, dimred="PCA")

plotTSNE(test, color_by = "cell_type_curated") +
    scale_color_manual(values = color_vector) +
    labs(color = "Ground truth")
p1 <- plotTSNE(test, color_by = "pr_label") +
    scale_color_manual(values = color_vector) +
    labs(color = "Point predictions")
p2 <- plotTSNE(test, color_by = "comm_anc_GI") +
    scale_color_manual(values = color_vector) +
    labs(color = "Graph-based labels")

library(ggpubr)
ggarrange(p1, p2, ncol=2, nrow=1)

```



